### PR TITLE
[quality] fix 'no-implicit-dependencies' errors

### LINKED ISSUE #5792: [quality] no implicit `@theia` dependencies
**Description**

In the past, we attempted to address violations of the **`no-implicit-dependencies`** tslint rule in Theia which warned us against declaring implicit dependencies. The problem with `@theia` implicit dependencies is that certain applications can build products with a subset of available extensions which may cause them issues if those extensions themselves have `@theia` implicit dependencies that the application does not declare (meaning the dependencies are not pulled by the application nor by other extensions).

Should there be a coding guideline to prevent the use of `@theia/` implicit dependencies since we cannot use an tslint `error` as we pick up non-theia offenders (ex: `vscode-jsonrpc`)

cc @akosyakov 

---

**Current Violators**

```bash
/node_modules/tslint/bin/tslint -p . | grep "dependency" | grep "@theia"
WARNING: /Users/vincentfugnitto/workspace/theia/dev-packages/cli/src/theia.ts:19:34 - Module '@theia/application-package' is not listed as dependency in package.json
WARNING: /Users/vincentfugnitto/workspace/theia/packages/debug/src/browser/console/debug-console-session.ts:25:133 - Module '@theia/languages' is not listed as dependency in package.json
WARNING: /Users/vincentfugnitto/workspace/theia/packages/filesystem/src/browser/download/file-download-command-contribution.ts:20:29 - Module '@theia/application-package' is not listed as dependency in package.json
WARNING: /Users/vincentfugnitto/workspace/theia/packages/filesystem/src/browser/filesystem-frontend-contribution.ts:19:29 - Module '@theia/application-package' is not listed as dependency in package.json
WARNING: /Users/vincentfugnitto/workspace/theia/packages/json/src/node/json-contribution.ts:19:29 - Module '@theia/application-package' is not listed as dependency in package.json
WARNING: /Users/vincentfugnitto/workspace/theia/packages/plugin-dev/src/browser/hosted-plugin-log-viewer.ts:18:53 - Module '@theia/output' is not listed as dependency in package.json
WARNING: /Users/vincentfugnitto/workspace/theia/packages/plugin-dev/src/browser/hosted-plugin-log-viewer.ts:19:36 - Module '@theia/output' is not listed as dependency in package.json
WARNING: /Users/vincentfugnitto/workspace/theia/packages/plugin-ext-vscode/src/browser/plugin-vscode-commands-contribution.ts:23:29 - Module '@theia/workspace' is not listed as dependency in package.json
WARNING: /Users/vincentfugnitto/workspace/theia/packages/plugin-ext-vscode/src/browser/plugin-vscode-commands-contribution.ts:24:31 - Module '@theia/editor' is not listed as dependency in package.json
WARNING: /Users/vincentfugnitto/workspace/theia/packages/plugin-ext-vscode/src/browser/plugin-vscode-commands-contribution.ts:31:35 - Module '@theia/workspace' is not listed as dependency in package.json
WARNING: /Users/vincentfugnitto/workspace/theia/packages/plugin-ext/src/main/browser/debug/debug-main.ts:40:33 - Module '@theia/terminal' is not listed as dependency in package.json
WARNING: /Users/vincentfugnitto/workspace/theia/packages/plugin-ext/src/main/browser/debug/debug-main.ts:42:38 - Module '@theia/output' is not listed as dependency in package.json
WARNING: /Users/vincentfugnitto/workspace/theia/packages/plugin-ext/src/main/browser/debug/plugin-debug-session-factory.ts:18:33 - Module '@theia/terminal' is not listed as dependency in package.json
WARNING: /Users/vincentfugnitto/workspace/theia/packages/plugin-ext/src/main/browser/debug/plugin-debug-session-factory.ts:23:38 - Module '@theia/output' is not listed as dependency in package.json
WARNING: /Users/vincentfugnitto/workspace/theia/packages/plugin-ext/src/main/browser/debug/plugin-debug-session-factory.ts:31:39 - Module '@theia/terminal' is not listed as dependency in package.json
WARNING: /Users/vincentfugnitto/workspace/theia/packages/plugin-ext/src/main/browser/language-provider/language-client-contribution-provider-impl.ts:23:44 - Module '@theia/languages' is not listed as dependency in package.json
WARNING: /Users/vincentfugnitto/workspace/theia/packages/plugin-ext/src/main/browser/language-provider/language-client-contribution-provider.ts:18:44 - Module '@theia/languages' is not listed as dependency in package.json
WARNING: /Users/vincentfugnitto/workspace/theia/packages/plugin-ext/src/main/browser/language-provider/plugin-language-client-provider.ts:18:40 - Module '@theia/languages' is not listed as dependency in package.json
WARNING: /Users/vincentfugnitto/workspace/theia/packages/plugin-ext/src/main/browser/language-provider/plugin-language-client-provider.ts:19:61 - Module '@theia/languages' is not listed as dependency in package.json
WARNING: /Users/vincentfugnitto/workspace/theia/packages/plugin-ext/src/main/browser/languages-contribution-main.ts:23:68 - Module '@theia/languages' is not listed as dependency in package.json
WARNING: /Users/vincentfugnitto/workspace/theia/packages/plugin-ext/src/main/browser/languages-contribution-main.ts:24:71 - Module '@theia/languages' is not listed as dependency in package.json
WARNING: /Users/vincentfugnitto/workspace/theia/packages/plugin-ext/src/main/browser/output-channel-registry-main.ts:18:30 - Module '@theia/output' is not listed as dependency in package.json
WARNING: /Users/vincentfugnitto/workspace/theia/packages/plugin-ext/src/main/browser/output-channel-registry-main.ts:19:36 - Module '@theia/output' is not listed as dependency in package.json
WARNING: /Users/vincentfugnitto/workspace/theia/packages/plugin-ext/src/main/browser/output-channel-registry-main.ts:20:53 - Module '@theia/output' is not listed as dependency in package.json
WARNING: /Users/vincentfugnitto/workspace/theia/packages/plugin-ext/src/main/browser/plugin-ext-frontend-module.ts:44:40 - Module '@theia/languages' is not listed as dependency in package.json
WARNING: /Users/vincentfugnitto/workspace/theia/packages/plugin-ext/src/main/browser/terminal-main.ts:20:32 - Module '@theia/terminal' is not listed as dependency in package.json
WARNING: /Users/vincentfugnitto/workspace/theia/packages/plugin-ext/src/main/browser/terminal-main.ts:21:33 - Module '@theia/terminal' is not listed as dependency in package.json
WARNING: /Users/vincentfugnitto/workspace/theia/packages/plugin-ext/src/plugin/languages/color.ts:19:34 - Module '@theia/plugin-ext' is not listed as dependency in package.json
WARNING: /Users/vincentfugnitto/workspace/theia/packages/plugin-ext/src/plugin/languages/color.ts:21:30 - Module '@theia/plugin-ext' is not listed as dependency in package.json
WARNING: /Users/vincentfugnitto/workspace/theia/packages/plugin-ext/src/plugin/languages/folding.ts:18:34 - Module '@theia/plugin-ext' is not listed as dependency in package.json
WARNING: /Users/vincentfugnitto/workspace/theia/packages/plugin-ext/src/plugin/languages/rename.ts:21:34 - Module '@theia/plugin-ext' is not listed as dependency in package.json
WARNING: /Users/vincentfugnitto/workspace/theia/packages/plugin-ext/src/plugin/languages/rename.ts:22:34 - Module '@theia/plugin-ext' is not listed as dependency in package.json
WARNING: /Users/vincentfugnitto/workspace/theia/packages/task/src/browser/task-service.ts:18:31 - Module '@theia/editor' is not listed as dependency in package.json
``` 

---

**Related Info**
- https://github.com/theia-ide/theia/pull/3754
- https://github.com/theia-ide/theia/pull/3044

### PR DESCRIPTION
**Pull request by vince-fugnitto**
Fixes #5792

- fixed existing `no-implicit-dependencies' tslint error
- updated the `no-implicit-dependencies` warning into an error
- added common `core` dependencies to the exsiting whitelist
- updated rule to only check for `dev` dependencies

Signed-off-by: Vincent Fugnitto <vincent.fugnitto@ericsson.com>

### CODE REVIEW

**Comment by marechal-p** [comment]
> to upgrade request version one has to update it in all packages

As long as the major version is the same, wouldn't the range allow yarn to only download one version? And if we don't care about the major version and we just want say core to fix it, then why not use `*` as a constraint in the other extensions? Maybe even as a peer dependency idk?

**Comment by vince-fugnitto** [comment]
@marcdumais-work removing the `dev` check means I get the following error:
```
$ theiaext build
[lint]
[lint] ERROR: /Users/vincentfugnitto/workspace/theia/packages/git/src/node/dugite-git.spec.ts:18:24 - Module 'upath' is not listed as dependency in package.json
[lint]
```

https://github.com/theia-ide/theia/blob/f5161dbc4882b800e928a11cf8b0901d76e29eee/packages/git/src/node/dugite-git.spec.ts#L18

It's listed as a `devDependency` in the `git` extension:

https://github.com/theia-ide/theia/blob/f5161dbc4882b800e928a11cf8b0901d76e29eee/packages/git/package.json#L67-L69

**Comment by akosyakov** [comment]
> As long as the major version is the same, wouldn't the range allow yarn to only download one version? And if we don't care about the major version and we just want say core to fix it, then why not use * as a constraint in the other extensions? Maybe even as a peer dependency idk?

@marechal-p I have doubts about it. yarn is not smart about it, there is no SAT resolver or something like that to find optimal dependency configuration which excludes duplicates as much as possible. Usually it deduplicates only latest versions allowed by ranged are pointing to the same version then it uses it, otherwise it installs both versions. So if newer version is released than one which is fixed in core, then `*` will lead to installing different versions in additional to aversion from the core.

You can experiment though. For now we have a script from @kittaakos which prints duplicate dependencies after install, we should pay attention to it and resolve them as we see them.

File: configs/errors.tslint.json [hunk 1]
```diff
@@ -38,6 +38,51 @@
...
+          "fs-extra",
+          "inversify",
+          "jsdom",
+          "lodash.debounce",
+          "lodash.throttle",
+          "monaco-languageclient",
+          "native-keymap",
+          "nsfw",
+          "react",
+          "react-autosize-textarea",
```
Comment (akosyakov) [inline configs/errors.tslint.json]: strange that we have it in core, but it is used only in scm, maybe we better just to declare it scm for now

**Comment by akosyakov** [comment]
https://github.com/theia-ide/theia/pull/5808#issuecomment-515568450

@vince-fugnitto I agree with @marcdumais-work that we should not allow importing dev dependencies in prod code. It's error-prone. For this case, can we use `disable-next-line`?

File: configs/errors.tslint.json (hunk 1 above)
Comment (vince-fugnitto) [reply configs/errors.tslint.json]: I'll make the changes :)

**Comment by vince-fugnitto** [comment]
@akosyakov is it fine to merge?

**Comment by akosyakov** [comment]
yes

**Pull request merged**